### PR TITLE
Rd 6780 delta node tracer truncated request s http body does not get scrubbed

### PR DESCRIPTION
**Pull request by OrrLumigo**

### CODE REVIEW

File: src/hooks/http.ts [hunk 1]
```diff
@@ -346,7 +346,9 @@ export class Http {
         if (isEmptyString(requestData.body)) {
           const body = extractBodyFromEmitSocketEvent(args[1]);
           if (body) {
-            requestData.body += body.substr(0, getEventEntitySize(false));
+            const eventEntitySize = getEventEntitySize(false);
+            requestData.body += body.substr(0, eventEntitySize);
+            requestData.truncated = eventEntitySize < requestData.body.length;
```
Comment (saartochner) [inline src/hooks/http.ts:351]:
```suggestion
            requestData.truncated = eventEntitySize < body.length;
```
I think that you're truncate only the body

Comment (saartochner) [inline src/hooks/http.ts:351]: Although I do agree that this code should be changed to:

```suggestion
            const newBody = requestData.body + body;
            requestData.body = newBody.substr(0, eventEntitySize);
            requestData.truncated = eventEntitySize < newBody.length;
```

Comment (saartochner) [reply src/hooks/http.ts:351]: please add tests to these two cases

Comment (OrrLumigo) [reply src/hooks/http.ts:351]: why ask on the newBody we should ask on the whole body collected so far

File: src/hooks/http.ts [hunk 2]
```diff
@@ -47,10 +47,9 @@ export class Http {
     return function (args) {
       GlobalDurationTimer.start();
       if (isEmptyString(requestData.body)) {
-        const body = extractBodyFromWriteOrEndFunc(args);
-        if (body) {
-          requestData.body += body.substr(0, getEventEntitySize(false));
-        }
+        const [body, truncated] = extractBodyFromWriteOrEndFunc(args);
+        if (body) requestData.body += body;
```
Comment (saartochner) [inline src/hooks/http.ts:51]:
```suggestion
        if (body) {
         requestData.body += body;
         if (requestData.body > getEventEntitySize(false)) {
             requestData.body = substr(requestData.body);
             truncated = true;
         }
   }
```

File: src/reporter.test.ts [hunk 3]
```diff
@@ -330,7 +332,61 @@ describe('reporter', () => {
       const actual = JSON.parse(reporter.forgeAndScrubRequestBody(spans, expectedResultSize));
       expect(actual).toEqual(expected);
     });
+    test('forgeAndScrubRequestBody long request', () => {
+      let manyA = 'a'.repeat(88);
+      let manyManyA = 'a'.repeat(1268);
```
Comment (saartochner) [inline src/reporter.test.ts:337]: what is this number? call the relevant function

File: src/reporter.test.ts [hunk 4]
```diff
@@ -330,7 +332,61 @@ describe('reporter', () => {
...
+        { dummyEnd },
+      ];
+      const expected = [
+        {
+          info: {
+            httpInfo: {
+              host: 'your.mind.com',
+              request: {
+                truncated: true,
+                body: `{"a":"${manyA}","b":"${manyA}","key":"****","password":"****","e":"${manyA}","secret":"****","f":"${manyA}","g":"${manyA}","h":"${manyManyA}"}`,
```
Comment (saartochner) [inline src/reporter.test.ts:371]: shouldn't you need to truncate the body? it is still very long 🤔

File: src/spans/awsSpan.ts [hunk 5]
```diff
@@ -238,10 +238,10 @@ export const decodeHttpBody = (httpBody: any, hasError: boolean): any | string =
   return httpBody;
 };
 
-export const getHttpInfo = (requestData, responseData, truncated = false): HttpInfo => {
+export const getHttpInfo = (requestData, responseData): HttpInfo => {
   const { host } = requestData;
   const request = Object.assign({}, requestData);
-  const response = Object.assign({ truncated }, responseData);
+  const response = Object.assign({}, responseData);
```
Comment (saartochner) [inline src/spans/awsSpan.ts:244]: I'm not sure what is the purpose of this line - is this a copy?

Comment (OrrLumigo) [reply src/spans/awsSpan.ts:244]: we moved the truncated field to another place.

File: src/hooks/http.ts [hunk 6]
```diff
@@ -48,15 +48,23 @@ export class Http {
...
   }
 
+  static aggregateRequestBodyToSpan(body, requestData, currentSpan) {
+    if (body && !requestData.truncated) {
+      const eventEntitySize = getEventEntitySize(true);
+      requestData.body += body;
+      const truncated = eventEntitySize < requestData.body.length;
+      if (truncated) requestData.body = requestData.body.substr(0, eventEntitySize);
+      requestData.truncated = truncated;
+      if (currentSpan) currentSpan.info.httpInfo = getHttpInfo(requestData, {});
```
Comment (saartochner) [inline src/hooks/http.ts:64]: this line should be executed even if the body is empty

File: src/hooks/http.ts [hunk 7]
```diff
@@ -48,15 +48,23 @@ export class Http {
...
-          requestData.body += body.substr(0, getEventEntitySize(false));
-        }
-        if (currentSpan) currentSpan.info.httpInfo = getHttpInfo(requestData, {});
+        Http.aggregateRequestBodyToSpan(body, requestData, currentSpan);
       }
       GlobalDurationTimer.stop();
     };
   }
 
+  static aggregateRequestBodyToSpan(body, requestData, currentSpan) {
```
Comment (saartochner) [inline src/hooks/http.ts:57]: please add tests to every aspect of this function.
It is extremely complex, so it should be tested deeply

**Review by saartochner — CHANGES_REQUESTED** [review]
GJ!

**Review by saartochner — APPROVED** [review]

File: src/hooks/http.test.js [hunk 8]
```diff
@@ -188,7 +280,7 @@ describe('http hook', () => {
     const wrapper = Http.httpRequestWriteBeforeHookWrapper(requestData);
     wrapper([firstArg, secArg]);
 
-    expect(requestData).toEqual({ body: 'BODY' });
```
Comment (doriaviram) [inline src/hooks/http.test.js:191]: Add some test with `truncated=true`

File: src/hooks/http.ts [hunk 9]
```diff
@@ -48,15 +48,27 @@ export class Http {
...
       }
       GlobalDurationTimer.stop();
     };
   }
 
+  static aggregateRequestBodyToSpan(
+    body,
+    requestData,
+    currentSpan,
+    maxSize = getEventEntitySize(true)
```
Comment (doriaviram) [inline src/hooks/http.ts:61]: Can we add types here ?

File: src/hooks/http.ts [hunk 10]
```diff
@@ -48,15 +48,27 @@ export class Http {
...
+  ) {
+    if (body && !requestData.truncated) {
+      requestData.body += body;
+      const truncated = maxSize < requestData.body.length;
+      if (truncated) requestData.body = requestData.body.substr(0, maxSize);
+      requestData.truncated = truncated;
+    }
+    if (currentSpan) currentSpan.info.httpInfo = getHttpInfo(requestData, {});
```
Comment (doriaviram) [inline src/hooks/http.ts:69]: Do we need to do `getHttpInfo` before the truncate?

Comment (doriaviram) [inline src/hooks/http.ts:69]: `getHttpInfo` get also `truncated` param, its will override your one ?

File: src/reporter.test.ts [hunk 11]
```diff
@@ -311,13 +311,124 @@ describe('reporter', () => {
...
+              request: {
+                truncated: false,
+                body: '{"secret":"****"}',
+                headers: '{"Tyler":"Durden","secretKey":"****","content-type":"application/json"}',
+                host: 'your.mind.com',
+              },
+              response: {
+                truncated: false,
+                body: `{"key":"****","password":"****","e":"${value}","secret":"****","f":"${value}","g":"${value}","h":"${shorter}"}...[too long]`,
+                headers: '{"Peter":"Parker","content-type":"application/json"}',
```
Comment (doriaviram) [inline src/reporter.test.ts:378]: This one looks like it's truncated, because of the `...[to long]`

Comment (OrrLumigo) [reply src/reporter.test.ts:378]: truncated not related to [too long] it indicates either we stoped collecting the body

File: src/hooks/http.test.js (hunk 8 above)
Comment (OrrLumigo) [reply src/hooks/http.test.js:191]: there are tests for this elsewhere in reporter.test

 test('forgeAndScrubRequestBody truncated response', () => {

File: src/hooks/http.ts (hunk 10 above)
Comment (OrrLumigo) [reply src/hooks/http.ts:69]: this function doesnt except this field anymore

Comment (OrrLumigo) [reply src/hooks/http.ts:69]: it was always outside the if
if (body && !requestData.truncated) {

File: src/hooks/http.ts (hunk 9 above)
Comment (OrrLumigo) [reply src/hooks/http.ts:61]: we dont have types yet for currentSpan

File: src/reporter.test.ts [hunk 12]
```diff
@@ -311,13 +311,124 @@ describe('reporter', () => {
...
+                truncated: false,
                 body: '{"secret":"****"}',
                 headers: '{"Tyler":"Durden","secretKey":"****","content-type":"application/json"}',
                 host: 'your.mind.com',
               },
               response: {
                 truncated: true,
-                body: `{"a":"${manyA}","b":"${manyA}","key":"****","password":"****","e":"${manyA}","secret":"****","f":"${manyA}","g":"${manyA}","h":"${manyManyA}"}`,
+                body: `{"a":"${value}","b":"${value}","key":"****","password":"****","e":"${value}","secret":"****","f":"${value}","g":"${value}"}`,
+                headers: '{"Peter":"Parker","content-type":"application/json"}',
```
Comment (doriaviram) [inline src/reporter.test.ts:322]: @OrrLumigo Its a bug, we need to fix that

**Pull request merged**